### PR TITLE
[Caffe2] Detectron fpn support

### PR DESCRIPTION
**Pull request by daquexian**
To [support detectron's fpn on cpu](https://github.com/facebookresearch/Detectron/issues/134), some layers in caffe2 need to be implemented.

 - [x] Update collect_and_distribute_fpn_rpn_proposals_op to fit arbitrary size (In existing implementation proposals from different fpn levels must have the same size, which is obviously unnecessary and unlikely. 
 - [x] Implement batch_permutation on cpu

### CODE REVIEW

File: modules/detectron/batch_permutation_op.h [hunk 1]
```diff
@@ -32,8 +33,37 @@ class BatchPermutationOp final : public Operator<Context> {
...
+    const int H = X.dim32(2);
+    const int W = X.dim32(3);
+
+    const float *src = X.template data<float>();
+    float *dst = Y->template mutable_data<float>();
+
+    for (int i = 0; i < N; i++) {
+      int idx = indices.template data<int>()[i];
+
+      memcpy(dst + i * C * H * W, src + idx * C * H * W, sizeof(float) * C * H * W);
```
Comment (BIT-silence) [inline modules/detectron/batch_permutation_op.h]: If you only want to implement it on CPU. please move the definition of RunOnDevice to cc file

File: modules/detectron/batch_permutation_op.h [hunk 2]
```diff
@@ -32,8 +33,37 @@ class BatchPermutationOp final : public Operator<Context> {
...
+      ")");
+
+    Y->ResizeLike(X);
+
+    const int N = X.dim32(0);
+    const int C = X.dim32(1);
+    const int H = X.dim32(2);
+    const int W = X.dim32(3);
+
+    const float *src = X.template data<float>();
```
Comment (BIT-silence) [inline modules/detectron/batch_permutation_op.h]: const T* src = X.template data<T>();

File: modules/detectron/batch_permutation_op.h [hunk 3]
```diff
@@ -32,8 +33,37 @@ class BatchPermutationOp final : public Operator<Context> {
...
+    float *dst = Y->template mutable_data<float>();
```
Comment (BIT-silence) [inline modules/detectron/batch_permutation_op.h]: T* dst = Y->template mutable_data<T>();

File: modules/detectron/batch_permutation_op.h [hunk 4]
```diff
@@ -32,8 +33,37 @@ class BatchPermutationOp final : public Operator<Context> {
   USE_OPERATOR_CONTEXT_FUNCTIONS;
 
   bool RunOnDevice() override {
-    // No CPU implementation for now
-    CAFFE_NOT_IMPLEMENTED;
+    const auto& X = Input(0);
+    auto& indices = Input(1);
+    auto* Y = Output(0);
+
+    CAFFE_ENFORCE(indices.ndim() == 1, "indices must be 1-d");
```
Comment (BIT-silence) [inline modules/detectron/batch_permutation_op.h]: Please use CAFFE_ENFORCE_EQ, then when it fails the error log will be more detailed.

File: caffe2/operators/collect_and_distribute_fpn_rpn_proposals_op.cc [hunk 5]
```diff
@@ -115,47 +115,40 @@ bool CollectAndDistributeFpnRpnProposalsOp<CPUContext>::RunOnDevice() {
...
+  int proposal_num = 0;
   for (int i = 0; i < num_rpn_lvls; i++) {
-    // roi_in: (N, 5)
     const auto& roi_in = Input(i);
-    CAFFE_ENFORCE_EQ(roi_in.dims(), (vector<TIndex>{N, 5}));
+    proposal_num += roi_in.dim(0);
+  }
+  ERArrXXf rois(proposal_num, 5);
+  EArrXf scores(proposal_num);
+  for (int i = 0, len = 0; i < num_rpn_lvls; i++) {
```
Comment (BIT-silence) [inline caffe2/operators/collect_and_distribute_fpn_rpn_proposals_op.cc]: Please put int len = 0; out of the for loop.

File: modules/detectron/batch_permutation_op.h [hunk 6]
```diff
@@ -32,8 +33,37 @@ class BatchPermutationOp final : public Operator<Context> {
   USE_OPERATOR_CONTEXT_FUNCTIONS;
 
   bool RunOnDevice() override {
-    // No CPU implementation for now
-    CAFFE_NOT_IMPLEMENTED;
+    const auto& X = Input(0);
+    auto& indices = Input(1);
```
Comment (BIT-silence) [inline modules/detectron/batch_permutation_op.h]: const auto& indices = Input(1);

File: modules/detectron/batch_permutation_op.h (hunk 1 above)
Comment (BIT-silence) [reply modules/detectron/batch_permutation_op.h]: std::memcpy(...)

File: modules/detectron/batch_permutation_op.h [hunk 7]
```diff
@@ -21,6 +21,7 @@
 #include "caffe2/core/logging.h"
 #include "caffe2/core/operator.h"
 #include "caffe2/utils/math.h"
+#include <cstring>
```
Comment (BIT-silence) [inline modules/detectron/batch_permutation_op.h]: Put the include files of in front of others. Please refer to https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: modules/detectron/batch_permutation_op.h (hunk 1 above)
Comment (BIT-silence) [reply modules/detectron/batch_permutation_op.h]: Here maybe we can consider to make it in-place.

**Comment by daquexian** [comment]
@BIT-silence Thanks for your review! I have made requested changes except make batch_permutation in-place. Should it be finished now?

**Comment by BIT-silence** [comment]
The in-place part is just one of my idea. We can make it later. Currently we can use it in copy which make the GPU things easier. Thanks!

**Review by xiaomengy — APPROVED** [review]

**Comment by ZesongLee** [comment]
How could I use the fpn support in python?

**Comment by daquexian** [comment]
Python is already fully supported in detectron.

**Pull request merged**